streams/src/main/java/org/apache/kafka/streams/processor/ProcessorContext.java

```diff
@@ -85,8 +85,23 @@
      * @throws IllegalStateException If store gets registered after initialized is already finished
      * @throws StreamsException if the store's change log does not contain the partition
      */
+    default void register(final StateStore store,
+                  final StateRestoreCallback stateRestoreCallback) {
+        register(store, stateRestoreCallback, null);
+    }
+
+    /**
+     * Register and possibly restores the specified storage engine.
+     *
+     * @param store the storage engine
+     * @param stateRestoreCallback the restoration callback logic for log-backed state stores upon restart
+     * @param checkpointCallback called to checkpoint position metatadata of state stores
+     *
+     * @throws IllegalStateException If store gets registered after initialized is already finished
+     * @throws StreamsException if the store's change log does not contain the partition
+     */
     void register(final StateStore store,
-                  final StateRestoreCallback stateRestoreCallback);
+                  final StateRestoreCallback stateRestoreCallback, final CheckpointCallback checkpointCallback);
```

Comment:
@guozhangwang many thanks for sharing those thoughts! I think tracking position on a per subtopology/task level is really interesting and worth giving it more thoughts. It would make things simpler and the code more robust, and the drawbacks might not actually that big of a problem. Pragmatically given the point we are with this PR I still suggest we finish the current approach first and then explore the per task position in a follow-up. 

Re checkpoint callback, I think generalizing the name is a good idea.

Thanks again for the great review!

Reply:
Thanks for that naming suggestion, @guozhangwang ! I agree that "commit" is a much better term than "checkpoint".

Regarding your larger point, I agree that there's a burden on store maintainers to track the position correctly and to integrate correctly with state restoration and checkpointing. I think that we can probably address the burden of this API without sacrificing the precise correctness of position tracking. As you mentioned, it'll be easier to consider an alternative as a follow-on PR instead of continuing to revise this one, though. I'll file a ticket shortly to circle back on this issue before the next release.

Reply:
I've filed a 3.2.0 blocker to revisit this: https://issues.apache.org/jira/browse/KAFKA-13622